Applications/Spire/Source/Ui/DropDownList.cpp

```diff
@@ -0,0 +1,259 @@
+        case Qt::Key_Down:
+          if(isVisible()) {
+            activate_next();
+            return true;
+          }
+          break;
+        case Qt::Key_Up:
+          if(isVisible()) {
+            activate_previous();
+            return true;
+          }
+          break;
+        case Qt::Key_Enter:
+        case Qt::Key_Return:
+          if(isVisible() && m_highlight_index) {
+            on_item_selected(get_widget(*m_highlight_index)->get_value(),
+              *m_highlight_index);
+            return true;
+          }
+          break;
+      }
+    }
+  }
+  return DropDownWindow::eventFilter(watched, event);
+}
+
+void DropDownList::hideEvent(QHideEvent* event) {
+  m_scroll_area->verticalScrollBar()->setValue(0);
+  if(m_highlight_index) {
+    auto item = m_layout->itemAt(*m_highlight_index);
+    if(item != nullptr) {
+      static_cast<DropDownItem*>(item->widget())->reset_highlight();
+    }
+    m_highlight_index = boost::none;
+  }
+}
+
+void DropDownList::keyPressEvent(QKeyEvent* event) {
+  switch(event->key()) {
+    case Qt::Key_Down:
+      if(isVisible()) {
+        activate_next();
+      }
+      break;
+    case Qt::Key_Enter:
+      if(m_highlight_index) {
+        on_item_selected(get_widget(*m_highlight_index)->get_value(),
+          *m_highlight_index);
+      }
+      break;
+    case Qt::Key_Escape:
+      hide();
+      break;
+    case Qt::Key_Up:
+      if(isVisible()) {
+        activate_previous();
+      }
+      break;
+  }
+  DropDownWindow::keyPressEvent(event);
+}
+
+connection DropDownList::connect_activated_signal(
+    const ActivatedSignal::slot_type& slot) const {
+  return m_activated_signal.connect(slot);
+}
+
+connection DropDownList::connect_highlighted_signal(
+    const HighlightedSignal::slot_type& slot) const {
+  return m_highlighted_signal.connect(slot);
+}
+
+connection DropDownList::connect_index_selected_signal(
+    const IndexSelectedSignal::slot_type& slot) const {
+  return m_index_selected_signal.connect(slot);
+}
+
+connection DropDownList::connect_value_selected_signal(
+    const ValueSelectedSignal::slot_type& slot) const {
+  return m_value_selected_signal.connect(slot);
+}
+
+QVariant DropDownList::get_value(int index) {
+  if(m_layout->count() - 1 < index) {
+    return QVariant();
+  }
+  return get_widget(index)->get_value();
+}
+
+void DropDownList::insert_item(DropDownItem* item) {
+  m_item_selected_connections.push_back(std::move(
+    item->connect_selected_signal(
+      [=, index = m_layout->count()] (auto value) {
+        on_item_selected(std::move(value), index);
+      })));
+  m_layout->insertWidget(m_layout->count(), item);
+  update_height();
+}
+
+int DropDownList::item_count() const {
+  return m_layout->count();
+}
+
+void DropDownList::remove_item(int index) {
+  if(index > m_layout->count() - 1) {
+    return;
+  }
+  m_item_selected_connections.pop_back();
+  for(auto i = std::size_t(index); i < m_item_selected_connections.size();
```

Comment:
Consider using `std::list` for `m_item_selected_connections`. List iterators won't be invalidated by insertions and deletions, so you may be able to avoid the array index trouble altogether.

Reply:
I'm not sure how iterators would simplify this, the API for both removing and signalling item selection is done with indexes.

Reply:
You don't have to use indexes in the implementation just because the API works with them.

The problem with the existing code is that, while list items (connections) do not depend on each other, you have to update all the items on every write operation because array indexes do not completely identify the locations of the elements as you move them around. The solution to this problem is to fix the locations of the elements, capture those fixed locations in your lambdas, and move around handles rather than the elements themselves.

For example, consider storing `std::unique_ptr<boost::signals2::scoped_connection>` instead of `boost::signals2::scoped_connection` directly in `m_item_selected_connections`. On line 132, you could then capture the fixed address of the connection using `std::unique_ptr::get` so that you would never need to update the signal function. Of course, you would still need to keep some code to entertain the index API, but it does not have to be as operationally complex.

Reply:
The connections are returned by value; wouldn't they get deleted when this function returns?